inst/examples/shared.cpp

```diff
@@ -0,0 +1,97 @@
+// Same as the arrays example but using shared memory. We split our
+// internal storage into two chunks;
+//
+// * one is "internal" and can be arbitrarily modified; this is the
+//   same as our normal internal state really and will be duplicated
+//   for each particle.
+// * the other is "shared" and will be stored as a "const reference";
+//   we will guarantee that Dust holds a copy of the 'init_t' object
+//   so that this reference stays valid.
+#include <array>
+#include <memory>
+
+class arrays {
+public:
+  typedef double real_t;
+
+  struct shared_t {
+    int dim_r;
+    int dim_x;
+    int n;
+    real_t initial_y;
+    std::array<real_t, 3> initial_x;
+  };
```

Comment:
Do we definitely only hold one copy of this in memory, as every particle will contain an `arrays` class, each of which has a shared_t struct? Should this be `static`?
Or am I misunderstanding?

Reply:
I must admit I never remember all the different ways that `static` can be used. But if it's for a struct or class member would that not shared by all instantiations of the class? Or are you meaning something else? (Where in the above are you thinking of putting `static`?)

Reply:
i.e.:
```
  static struct shared_t {
    int dim_r;
    int dim_x;
    int n;
    real_t initial_y;
    std::array<real_t, 3> initial_x;
  };
```
So that it is shared by all instantiations of the class - isn't that what we want?

Reply:
It's not though - if two copies of this are created with different `n` then they'll have different values. It's the same for all particles within a dust object though

Reply:
Right, yes, that makes sense - no static then.
I'm getting in a muddle between struct definition and instance I think. `dust_data` actually creates this struct and the pointer, and is only called once?

Reply:
yeah, this is just the definition of the struct which basically namespaces it for us

Reply:
Right, I will merge then!